src/test/java/org/stellar/sdk/Sep10ChallengeTest.java

```diff
@@ -46,7 +50,1296 @@ public void testChallenge() throws IOException {
+            timeBounds
+    );
+
+    Sep10Challenge.ChallengeTransaction challengeTransaction = Sep10Challenge.readChallengeTransaction(challenge, server.getAccountId(), Network.TESTNET);
+    assertEquals(challengeTransaction, new Sep10Challenge.ChallengeTransaction(Transaction.fromEnvelopeXdr(challenge, Network.TESTNET), client.getAccountId()));
+  }
+
+  @Test
+  public void testReadChallengeTransactionValidSignedByServer() throws InvalidSep10ChallengeException, IOException {
+    KeyPair server = KeyPair.random();
+    KeyPair client = KeyPair.random();
+    Network network = Network.TESTNET;
+
+    long now = System.currentTimeMillis() / 1000L;
+    long end = now + 300;
+    TimeBounds timeBounds = new TimeBounds(now, end);
+
+    String challenge = Sep10Challenge.newChallenge(
+            server,
+            network,
+            client.getAccountId(),
+            "Stellar Test",
+            timeBounds
+    );
+
+    Transaction transaction = Transaction.fromEnvelopeXdr(challenge, network);
+    transaction.sign(client);
+
+    Sep10Challenge.ChallengeTransaction challengeTransaction = Sep10Challenge.readChallengeTransaction(challenge, server.getAccountId(), Network.TESTNET);
+    assertEquals(challengeTransaction, new Sep10Challenge.ChallengeTransaction(Transaction.fromEnvelopeXdr(challenge, Network.TESTNET), client.getAccountId()));
+  }
+
+  @Test
+  public void testReadChallengeTransactionInvalidNotSignedByServer() throws IOException {
+    KeyPair server = KeyPair.random();
+    KeyPair client = KeyPair.random();
+    String anchorName = "Stellar Test";
+    Network network = Network.TESTNET;
+
+    long now = System.currentTimeMillis() / 1000L;
+    long end = now + 300;
+    TimeBounds timeBounds = new TimeBounds(now, end);
+
+    byte[] nonce = new byte[48];
+    SecureRandom random = new SecureRandom();
+    random.nextBytes(nonce);
+    BaseEncoding base64Encoding = BaseEncoding.base64();
+    byte[] encodedNonce = base64Encoding.encode(nonce).getBytes();
+
+    Account sourceAccount = new Account(server.getAccountId(), -1L);
+    ManageDataOperation manageDataOperation1 = new ManageDataOperation.Builder(anchorName + " auth", encodedNonce)
+            .setSourceAccount(client.getAccountId())
+            .build();
+
+    Operation[] operations = new Operation[]{manageDataOperation1};
+    Transaction transaction = new Transaction(
+            sourceAccount.getAccountId(),
+            100 * operations.length,
+            sourceAccount.getIncrementedSequenceNumber(),
+            operations,
+            Memo.none(),
+            timeBounds,
+            network
+    );
+
+    transaction.sign(client);
+    try {
+      Sep10Challenge.readChallengeTransaction(transaction.toEnvelopeXdrBase64(), server.getAccountId(), network);
+      fail();
+    } catch (InvalidSep10ChallengeException e) {
+      assertEquals(String.format("Transaction not signed by server: %s.", server.getAccountId()), e.getMessage());
+    }
+  }
+
```

Comment:
❗️ Could we add a test here that tests what happens if the transaction base64 that is passed into the `readChallengeTransaction` function is corrupted?

This is the `TestReadChallengeTx_invalidCorrupted` test in the Go SDK:
https://github.com/stellar/go/blob/f4e38b5/txnbuild/transaction_test.go#L1253-L1281

Reply:
https://github.com/stellar/java-stellar-sdk/blob/dbd69a425b0ac43ce2f95c1e8d2948f2bc095449/src/test/java/org/stellar/sdk/Sep10ChallengeTest.java#L146L172

I think we should throw an exception that fails to read XDR in this [function](https://github.com/stellar/java-stellar-sdk/blob/dbd69a425b/src/main/java/org/stellar/sdk/Transaction.java#L192), so that we can easily catch it elsewhere, just like this: https://github.com/stellar/java-stellar-sdk/pull/264#discussion_r381269520

Reply:
Does the function currently not throw an exception trying to decode invalid XDR?

Reply:
As long as the function throws an exception I think you can just catch that here, and we can do any improvement to that in a separate PR.

Reply:
If we can do this: https://github.com/stellar/java-stellar-sdk/pull/264/files/e82649ae8e3b84c159aa7395be8b487177328bdc..dbd69a425b0ac43ce2f95c1e8d2948f2bc095449#r381556847 I think we're good.